code/datums/components/forensics.dm

```diff
@@ -0,0 +1,169 @@
+		fibers[i] = i
+	return TRUE
+
+/datum/component/forensics/proc/add_fibers(mob/living/carbon/human/M)
+	var/fibertext
+	var/item_multiplier = isitem(src)?1.2:1
+	if(M.wear_suit)
+		fibertext = "Material from \a [M.wear_suit]."
+		if(prob(10*item_multiplier) && !LAZYACCESS(fibers, fibertext))
+			LAZYSET(fibers, fibertext, fibertext)
+		if(!(M.wear_suit.body_parts_covered & CHEST))
+			if(M.w_uniform)
+				fibertext = "Fibers from \a [M.w_uniform]."
+				if(prob(12*item_multiplier) && !LAZYACCESS(fibers, fibertext)) //Wearing a suit means less of the uniform exposed.
+					LAZYSET(fibers, fibertext, fibertext)
+		if(!(M.wear_suit.body_parts_covered & HANDS))
+			if(M.gloves)
+				fibertext = "Material from a pair of [M.gloves.name]."
+				if(prob(20*item_multiplier) && !LAZYACCESS(fibers, fibertext))
+					LAZYSET(fibers, fibertext, fibertext)
+	else if(M.w_uniform)
+		fibertext = "Fibers from \a [M.w_uniform]."
+		if(prob(15*item_multiplier) && !LAZYACCESS(fibers, fibertext))
+			// "Added fibertext: [fibertext]"
+			LAZYSET(fibers, fibertext, fibertext)
+		if(M.gloves)
+			fibertext = "Material from a pair of [M.gloves.name]."
+			if(prob(20*item_multiplier) && !LAZYACCESS(fibers, fibertext))
+				LAZYSET(fibers, fibertext, fibertext)
+	else if(M.gloves)
+		fibertext = "Material from a pair of [M.gloves.name]."
+		if(prob(20*item_multiplier) && !LAZYACCESS(fibers, fibertext))
+			LAZYSET(fibers, fibertext, fibertext)
+	return TRUE
+
+/datum/component/forensics/proc/add_hiddenprint_list(list/_hiddenprints)	//list(ckey = text)
+	if(!length(_hiddenprints))
+		return
+	LAZYINITLIST(hiddenprints)
+	for(var/i in _hiddenprints)	//We use an associative list, make sure we don't just merge a non-associative list into ours.
+		hiddenprints[i] = _hiddenprints[i]
+	return TRUE
+
+/datum/component/forensics/proc/add_hiddenprint(mob/living/M)
+	if(!M || !M.key)
+		return
+	var/hasgloves = ""
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		if(H.gloves)
+			hasgloves = "(gloves)"
+	var/current_time = time_stamp()
+	if(!LAZYACCESS(hiddenprints, M.key))
+		LAZYSET(hiddenprints, M.key, "First: [M.real_name]\[[current_time]\][hasgloves]. Ckey: [M.ckey]")
+	else
+		var/laststamppos = findtext(LAZYACCESS(hiddenprints, M.key), " Last: ")
+		if(laststamppos)
+			LAZYSET(hiddenprints, M.key, copytext(hiddenprints[M.key], 1, laststamppos))
+		hiddenprints[M.key] += " Last: [M.real_name]\[[current_time]\][hasgloves]. Ckey: [M.ckey]"	//made sure to be existing by if(!LAZYACCESS);else
+	fingerprintslast = M.ckey
+	return TRUE
+
+/datum/component/forensics/proc/add_blood_DNA(list/dna)		//list(dna_enzymes = type)
+	if(!length(dna))
+		return
+	LAZYINITLIST(blood_DNA)
+	for(var/i in dna)
+		blood_DNA[i] = dna[i]
+	return TRUE
+
+/datum/component/forensics/proc/signal_check_evidence()
+	. = NONE
+	if(length(fingerprints))
+		. |= FORENSICS_COMPONENT_HAS_FINGERPRINTS

```

Comment:
@Cyberboss This sort of signal feels indicative of a bigger problem with the overall implementation but I'm unsure if it's a problem with @kevinz000 just doing it wrong or if it's a sign that a component is the wrong way to go instead of a plain datum on the atom path.

I'm leaning slightly more toward the latter

Reply:
@Cyberboss